Allow to lock SQLite database during backup

agners opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
This implements two new Websocket API endpoints `backup/start` and `backup/end`. Those are meant to notify the Core when the Supervisor is creating a backup of the Core. Since backups are done while the Core is running, this allows the Core to make sure all files in the `config` directory are in a safe state to read. Until now, the recorder continued to write to the database file while the backup is being made. This can lead to corrupted database files in the backup. With this implementation, the recorder locks the database using `BEGIN IMMEDIATE`, which essentially prevents writes from any Database connection while keeping the database available for reads. The implementation also halts the recorders queue so that no database inserts are even attempted.

This method is described as "historical" method in the SQLite documentation. Developers also confirmed that this method is still save to use and also works with WAL enabled.
https://www.sqlite.org/backup.html
https://sqlite.org/forum/forumpost/1307e369863b7b24?t=h

Note: Currently the `backup/start` and `backup/end` endpoints are implemented in the recorder directly. In case other components need to know when a backup is created, the Websocket API endpoint should be implemented as part of the `hassio` component or similar.


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [ ] There is no commented out code in this PR.
- [ ] I have followed the [development checklist][dev-checklist]
- [ ] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

agners committed:
Introduce Websocket API which allows to lock database during backup

agners committed:
Allow to set CONF_DB_URL

This is useful for test which need a custom DB path.

agners committed:
Introduce write_lock_db helper to lock SQLite database

agners committed:
Fix isort

bdraco commented on homeassistant/components/recorder/__init__.py:
```diff
@@ -788,34 +802,56 @@ def _run_external_statistics(self, metadata, stats):
+            self.queue.qsize(),
+        )
+
+    def _process_one_task(self, event) -> bool:
```
For a future PR, this linear `isinstance` search should probably be a `dict` lookup

agners replied:
Yeah or using a Task wrapper class for `Event` too, and introduce a base class with a handler funciton or similar. The reason I split it in this PR was a linter error. Agreed, optimizing it should be in a future PR.

bdraco commented on homeassistant/components/recorder/__init__.py:
```diff
@@ -982,6 +1018,41 @@ def block_till_done(self):
+            _LOGGER.warning("Database already locked.")
+            return False
+
+        task = DatabaseLockTask(threading.Event(), threading.Event(), False)
```
We could make it wait for an `asyncio.Event` and await that instead of typing up an executor thread

agners replied:
`asyncio.Event` is not thread safe, and afaict we are running in two different threads...

bdraco replied:
Rough implementation (untested) that avoids holding executor thread:
https://github.com/bdraco/home-assistant/commit/3f8a46c6f595830c00f7d1664d6a84387b7b77b8

agners replied:
Hm, I see that way around. Yeah seems safe, and avoids `run_in_executor`. I'll test it and add it to this PR.

emontnemery commented on tests/common.py:
```diff
@@ -897,10 +897,10 @@ def init_recorder_component(hass, add_config=None):
 
 
-async def async_init_recorder_component(hass, add_config=None):
+async def async_init_recorder_component(hass, config={}):
```
Let's not have mutable default arguments
```suggestion
async def async_init_recorder_component(hass, config=None):
```

agners replied:
Ok, I guess then we can leave the name as it was, essentially reverting this line:
```suggestion
async def async_init_recorder_component(hass, add_config=None):
```

emontnemery commented on tests/common.py:
```diff
@@ -897,10 +897,10 @@ def init_recorder_component(hass, add_config=None):
     """Initialize the recorder asynchronously."""
-    config = dict(add_config) if add_config else {}
-    config[recorder.CONF_DB_URL] = "sqlite://"
+    if recorder.CONF_DB_URL not in config:
+        config[recorder.CONF_DB_URL] = "sqlite://"
```
```suggestion
    config = config or {}
    if recorder.CONF_DB_URL not in config:
        config[recorder.CONF_DB_URL] = "sqlite://"
```

agners replied:
```suggestion
    config = add_config or {}
    if recorder.CONF_DB_URL not in config:
        config[recorder.CONF_DB_URL] = "sqlite://"
```

balloob commented:
Not to be merged before 2021.12 beta.

bdraco committed:
Avoid holding executor thread

agners committed:
Avoid mutable default arguments

agners committed:
Address pylint issues

emontnemery reviewed (commented):
We should add some negative test cases to make sure the error handling works.

agners committed:
Set unlock event in case timeout occures

This makes sure the database is left unlocked even in case of a race
condition.

agners committed:
Add more unit tests

agners commented:
> Not to be merged before 2021.12 beta.

As it will help to avoid corrupted backups I kinda was hoping that it still makes it in :pleading_face: 

Oth, it was the case for quite a while already :man_shrugging:

agners committed:
Address new pylint errors

emontnemery commented on tests/components/recorder/test_init.py:
```diff
@@ -1134,3 +1135,46 @@ def test_entity_id_filter(hass_recorder):
+
+    instance: Recorder = hass.data[DATA_INSTANCE]
+    with pytest.raises(TimeoutError):
+        await instance.lock_database(lock_timeout=1)
```
We can probably do a shorter timeout in tests?

agners replied:
My worry is that the test might end up to timeout for other reasons than the database being locked (e.g. not all events got processed within the time). In the end this kind of timeout testing are somewhat racy by nature to some degree.

A quick test on my somewhat old laptop showed that it normally takes about 10ms to lock the database. I'll set it to 100ms which should be fairly save.

emontnemery commented on homeassistant/components/recorder/__init__.py:
```diff
@@ -788,34 +802,60 @@ def _run_external_statistics(self, metadata, stats):
+        with write_lock_db(self):
+            # Notify that lock is being held, wait until database can be used again.
+            self.hass.add_job(_async_set_database_locked, task)
+            while not task.database_unlock.wait(timeout=1):
```
Move magic number to a constant to allow us to patch it in tests.

agners committed:
Lower timeout to speedup tests

agners committed:
Introduce queue overflow test

agners committed:
Unlock database if necessary

This makes sure that the test runs through in case locking actually
succeeds (and the test fails).

agners committed:
Make DB_LOCK_TIMEOUT a global

There is no good reason for this to be an argument. The recorder needs
to pick a sensible value.

agners committed:
Add Websocket Timeout test

agners committed:
Test lock_database() return

bdraco commented on homeassistant/components/recorder/__init__.py:
```diff
@@ -788,34 +805,60 @@ def _run_external_statistics(self, metadata, stats):
+            # Notify that lock is being held, wait until database can be used again.
+            self.hass.add_job(_async_set_database_locked, task)
+            while not task.database_unlock.wait(timeout=DB_LOCK_QUEUE_CHECK_TIMEOUT):
+                if self.queue.qsize() > MAX_QUEUE_BACKLOG * 0.9:
```
```suggestion
                if self.queue.qsize() > MAX_QUEUE_BACKLOG * 0.9:
```

Maybe make `MAX_QUEUE_BACKLOG * 0.9`  a constant?

agners replied:
Since the message just below refers to the 90% I like to have that factor nearby.

emontnemery commented on homeassistant/components/recorder/__init__.py:
```diff
@@ -687,6 +702,8 @@ def _run_event_loop(self):
     def _process_one_event_or_recover(self, event):
         """Process an event, reconnect, or recover a malformed database."""
         try:
+            if self._process_one_task(event):
+                return
             self._process_one_event(event)
```
The typical case is to record events, not handle special tasks.
Should we try that first?

agners replied:
No we can't as that code does not check the type. It accesses `event.event_type`, which is not a valid property on tasks.

Btw, the only reason I separated this in a function was because a linter complained about function lengths (the two lines seemed to pushed it over a limit).

emontnemery replied:
Implemented in https://github.com/home-assistant/core/pull/61161

emontnemery commented on homeassistant/components/recorder/__init__.py:
```diff
@@ -788,34 +805,60 @@ def _run_external_statistics(self, metadata, stats):
+            while not task.database_unlock.wait(timeout=DB_LOCK_QUEUE_CHECK_TIMEOUT):
+                if self.queue.qsize() > MAX_QUEUE_BACKLOG * 0.9:
+                    _LOGGER.warning(
+                        "Database queue backlog reached more than 90% of maximum queue length. Continue writing to database. Your Backup might be corrupted"
```
```suggestion
                        "Database queue backlog reached more than 90% of maximum queue "
						"length while waiting for backup to finish; recorder will now "
						"resume writing to database. The backup can not be trusted and "
						"must be restarted"
```

emontnemery reviewed (approved):
Looks great!

agners committed:
Update homeassistant/components/recorder/__init__.py

Co-authored-by: Erik Montnemery <erik@montnemery.com>

emontnemery committed:
Fix format

balloob commented on homeassistant/components/recorder/websocket_api.py:
```diff
@@ -106,3 +111,38 @@ def ws_info(
+@websocket_api.require_admin
+@websocket_api.websocket_command({vol.Required("type"): "backup/end"})
+@websocket_api.async_response
+async def ws_backup_end(
```
I know I am a little late to the party, but I thought of some improvements:

- Backup logic should live in the `hassio` integration. Recorder (and potentially in the future others) should indicate they have interest in this. Maybe via a `hassio.py` platform.
- WS API should be in hassio too.

And here for a bit more crazier one:
- If the supervisor were to disconnect abruptly, it would never call `backup_end`. That will leave the system in a broken state. It would be better if the backup_start would store the "end backup" logic in [`connection.subscriptions`](https://github.com/home-assistant/core/blob/dev/homeassistant/components/websocket_api/commands.py#L114-L118). Subscriptions are automatically finished when the connection is disconnected. That way, if the supervisor's websocket connection breaks, the backup mode is ended. The supervisor can then [end the subscription](https://developers.home-assistant.io/docs/api/websocket#unsubscribing-from-events) when it's done to also end the in-backup mode.

agners replied:
> * Backup logic should live in the hassio integration. Recorder (and potentially in the future others) should indicate they have interest in this. Maybe via a hassio.py platform.
> * WS API should be in hassio too.

Agreed, also suggested that in the PR intro (at least the WS part). At that point there was no other user, and I got no push back :angel: 

> If the supervisor were to disconnect abruptly, it would never call backup_end. That will leave the system in a broken state.

It would queue up events up to `MAX_QUEUE_BACKLOG * 0.9`, and then release the lock.

> It would be better if the backup_start would store the "end backup" logic in connection.subscriptions. 

Wouldn't that mean a single long running WS call from Supervisor side? That doesn't work, as I need to wait until the database is really locked (so `backup_start` not only notifies Core that backup is about to start, its return also signals the Supervisor that the Core is in fact ready to get going...).

balloob replied:
A subscription can send data over it to the caller. 

The Supervisor keeps it's WS API open the whole time I think.

emontnemery replied:
> Backup logic should live in the hassio integration. Recorder (and potentially in the future others) should indicate they have interest in this. Maybe via a hassio.py platform.

This seems backwards IMHO, the recorder does not care about hassio?
